### ISSUE TITLE
Note on the API design that involves code auto-generation

### ISSUE THREAD

**Issue by pepyakin** [paritytech/substrate#12331]
One of the best things about Rust is the balance between explicit and implicit things. See [here](https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html) to get the mental framework here. One example I want to highlight is importing things.

If you worked with languages like C, to print something, you would write `#include <stdio.h>`, and it would pull `printf` among other things. This can be very convenient for somebody who already knows C very well. They already know that `printf` comes from `stdio`. In other languages, such as Java or Rust, you have to manually specify everything you import with the fully qualified path, one by one. Additionally, those languages support the wildcard import, `*`, but it is frowned upon.

I consider that Rust made here a very good decision. It might be annoying to list every possible import, but you get a lot more in return. Let's turn the example upside down. Imagine you are reading code and seeing a previously unseen identifier. In Rust, you would need to grep within the file that identifier, and chances are you will find either an import or the declaration[^1]. Compare that to C: you will have to check every included header. Moreover, headers can include other headers as well. At that point, you will have to resort to a project-wide grep. C makes it easier to write code, and Rust optimizes for reading.

[^1]: Yes, I know there are exceptions, like inherent methods or traits. Or maybe when IDE is used. But bear with me, I still believe the point stands.

Now, how this translates to the API design? We have some macros that, unfortunately, do not follow the same spirit of API design. The worst offender so far is the runtime API.

Take for example the declaration of `apply_extrinsic`:

https://github.com/paritytech/substrate/blob/bd2a876e4c0e79b6e2ab6b0a92c038304a64b20f/primitives/block-builder/src/lib.rs#L27-L33

This will generate code that will implement this trait by calling into the Substrate Runtime at the current block ([apply_extrinsic](https://crates.parity.io/sp_block_builder/trait.BlockBuilder.html#method.apply_extrinsic)). Besides that it will generate a function ([apply_extrinsic_with_context](https://crates.parity.io/sp_block_builder/trait.BlockBuilder.html#method.apply_extrinsic_with_context)) for calling that Runtime API at a specific context, mostly used to call it at a state root other than the last best. 

Besides that, I am aware that there are functions that allow calling the Runtime API of a previous version. Here is the example,

https://github.com/paritytech/substrate/blob/bd2a876e4c0e79b6e2ab6b0a92c038304a64b20f/client/block-builder/src/lib.rs#L212-L218

Now, imagine a person reading this code with fresh eyes. If they are not aware of this code generation, and they try to grep `fn apply_extrinsic_before_version_6_with_context` they will end up with nothing and will end up very confusing.

I propose the following guideline: avoid pulling new identifiers from thin air. Generally, try to avoid changing the identifiers or signatures provided by the user. 

In particular, I think the following is the good shape of the API:

```rust
api.apply_extrinsic(xt.clone()) // Returns a struct a-la builder pattern.
    .at(block_id) 
    .context(ExecutionContext::BlockConstruction)
    .call();
```

It does not matter much how those are grouped, you can imagine `at` and `context` to be one call, and/or they can be the final calls (i.e. instead of the separate `call` call). That's not the point. The point is, that the `apply_extrinsic` has the same name as `BlockBuilder::apply_extrinsic`.

If you continue along those lines, you can arrive at an even better fitting API:

```rust
api.
    .at(block_id)
    .context(ExecutionContext::BlockConstruction)
    .apply_extrinsic(xt.clone());
```

This shape of the API has the same name and the same signature. Unfortunately, it may be less pragmatic to implement since `at` coming from multiple implementations will collide, but still may be worth pursuing.

**Comment by shawntabrizi** [comment]
cc @KiChjang @bkchr @sam0x17

**Comment by sam0x17** [comment]
I agree that this is a good place to use a builder pattern and I really like the syntax. One key requirement when doing something like this is to make sure it is well documented. This can be tricky with builder pattern because you absolutely need concrete examples since the normal way people poke around (seeing what methods exist on a thing) is more difficult with the builder pattern. So yeah I really like this as long as there will be some glaring documentation examples for the syntax that someone will see if they are poking around with the `api` object.

**Comment by pepyakin** [comment]
Just to clarify, I did not mean to fix only this particular macro. The builder pattern is also not as important, as my main point not to conjure identifiers out of thin air. I do not exclude that there could be other approaches, but that is also not the point of this post.

To address your comment, though, I agree that when you split a single API into a set of APIs, which is the case with the builder pattern, then it will be slightly trickier to understand. 

I don't think it's the end of the world, though: when you write runtime API unless you are directed to do that explicitly (somebody metaphorically types with your fingers), you have very likely seen some examples of how to deal with the particular API. Therefore, you probably know that there is the `context` followed by the `at` call.

Now, how they interact and what they do would not be particularly clear IMO. I think a good and sufficient solution would be to cross-link every part of the picture in the docs to a single place, and then explain there what is happening.

One potential way to do this, is to leave some documentation on the generated trait, that explains how to use it briefly, with an example, mentioning all kinds of different caveats. One caveat that could be mentioned is in the example below:

```
api.apply_extrinsic(xt.clone())
    .at(block_id) 
    .context(ExecutionContext::BlockConstruction)
    .call();
```

No work will be done in case `.call` is not ultimately called.

But TBH, I don't think it's worth discussing. IMO APIs at this level of user-facing-ness should have this kind of documentation by default.

**Comment by bkchr** [comment]
I thought about this already since quite some time. Aka that this `_context` and `at` parameter were not my best idea. I also thought that we should do a builder pattern, but in reverse. So, something like:

```
api.runtime_api(block_id).context(context).apply_exitrinsic()
```

Then there is also no problem around the documentation as there would be visible from the a normal trait.

**Comment by KiChjang** [comment]
I do believe that this is the direction that we were supposed to move towards (i.e. no magical identifiers coming out of nowhere), but the one macro that gives me pause is `construct_runtime`. It currently spits out quite a lot of identifiers during macro expansion, such as `RuntimeCall` and `RuntimeEvent`. We can't use the builder pattern here for this macro, so I'm wondering what should be the solution for macros that generate new types such as those mentioned above.

**Comment by bkchr** [comment]
You would need to change the syntax to make this more obvious. However, I don't have any idea about a good syntax.

**Comment by bkchr** [comment]
> I think the main block in turning a lot of the FRAME primitives to builders patters (which are arguably a lot nicer) is the over-use of **types**. Everything in FRAME is _types_, not a **value**, and types must be known at compile-time. Of course, the benefit is that we are avoiding a lot of dynamic dispatches by using types rather than values for calls and such.

How should this even look like? I mean what you are proposing there is a completely different DSL. One of the main points we are only operating on types is that the state is the storage and we don't use the memory to store any state. I also don't really get what your builder pattern should look like here? What would it do? If you really want to have some builder pattern, we could also make it work on the type level.
